sys/timer.c

```diff
@@ -0,0 +1,113 @@
+#define KL_LOG KL_TIME
+#include <timer.h>
+#include <mutex.h>
+#include <errno.h>
+#include <klog.h>
+#include <interrupt.h>
+
+static mtx_t timers_mtx = MTX_INITIALIZER(MTX_DEF);
+static timer_list_t timers = TAILQ_HEAD_INITIALIZER(timers);
+
+#define TMF_ACTIVE 0x1000
+#define TMF_INITIALIZED 0x2000
+#define TMF_ALLOCATED 0x4000
+#define TMF_REGISTERED 0x8000
+
+#define is_active(tm) ((tm)->tm_flags & TMF_ACTIVE)
+#define is_initialized(tm) ((tm)->tm_flags & TMF_INITIALIZED)
+#define is_allocated(tm) ((tm)->tm_flags & TMF_ALLOCATED)
+#define is_registered(tm) ((tm)->tm_flags & TMF_REGISTERED)
```

Comment:
Please describe in a comment how the state of a timer_t object changes in time, i.e first it is registered, next allocated then initialized. Finally it can transit forth and back between initialized and active states.

Reply:
I added inline [dot](https://www.graphviz.org/doc/info/lang.html) graph to explain it.